### PR TITLE
project_loader: process parts in consistent order

### PR DESCRIPTION
**Pull request by kyrofa**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [x] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh unit`?

-----

Currently, if parts are not strictly ordered by using `after`, the snapcraft CLI will process them in a completely non-deterministic order that changes between runs. This can make it extremely difficult to track down bugs that only occur when building in a specific order. This PR resolves [LP: #1759875](https://bugs.launchpad.net/snapcraft/+bug/1759875) by processing parts consistently and deterministically by sorting them by name.

### CODE REVIEW

File: tests/unit/project_loader/test_parts.py [hunk 1]
```diff
@@ -138,3 +138,71 @@ def test_load_with_dummy_repo(self, os_release_mock):
...
+                description: test
+                confinement: strict
+
+                parts:
+                  part2:
+                    plugin: nil
+                    after: [part3]
+                  part1:
+                    plugin: nil
+                    after: [part3]
```
Comment (sergiusens) [inline tests/unit/project_loader/test_parts.py]: can I see a test with only `part2` having `after: [part3]`?

Comment (kyrofa) [reply tests/unit/project_loader/test_parts.py]: Of course, done.

**Comment by kyrofa** [comment]
> I think we need one more test, to make sure consistency is a given with the use of after.

Done.

> Also, a counter proposal, albeit harder to implement would be to just use the ordered load from the OrderedDict and when an unsatisfied prereq is detected call it out in the order that it is declared in on the after list.

The YAML spec does not guarantee order in this case; I feel this would be relying on implementation details. I also like that the current approach will always process using the same order, regardless of the order things are declared.

**Comment by sergiusens** [comment]
@kyrofa I am aware of the spec not guaranteeing order, but we are providing one regardless.

**Comment by kyrofa** [comment]
I mean we aren't guaranteed to get them in order from pyyaml, so I'd rather not rely on it.

**Pull request merged**